libzecale/circuits/pairing/bls12_377_pairing.hpp

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef ZECALE_CIRCUITS_PAIRING_BLS12_377_PAIRING_HPP__
+#define ZECALE_CIRCUITS_PAIRING_BLS12_377_PAIRING_HPP__
+
+#include "libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp"
+
+#include <libff/algebra/curves/bls12_377/bls12_377_pp.hpp>
+#include <libsnark/gadgetlib1/gadgets/fields/fp2_gadgets.hpp>
+#include <libsnark/gadgetlib1/gadgets/pairing/pairing_params.hpp>
+
+namespace libzecale
+{
+
+/// Holds an element of G2 in homogeneous projective form. Used for
+/// intermediate values of R in the miller loop.
+template<typename ppT> class bls12_377_G2_proj
```

Comment:
Shouldn't this class inherit from `libsnark::gadget`?

i.e.
```c++
template<typename ppT>
class bls12_377_G2_proj : public libsnark::gadget<libff::Fr<ppT> > {
...
}
```
?

Reply:
It's intended as just a holder of variables, similar to `libsnark::G1_precomputation` and others, for example.  So I may be missing something, but I don't think this needs to be a gadget.

Reply:
I see, I find it confusing to have a method `generate_r1cs_witness` in his class which isn't a gadget. Do you really need it here?

Reply:
It's required to set the value, following the convention of the underlying Fqe_variable type and calls the method of exactly the same name on those variables. I didn't see an issue with a non-gadget having that method, but we can change it if you prefer.

I'm not sure it would be an improvement to introduce unnecessary inheritance, but the current name seems to be most fitting, given what it does. If you have a preference or a better idea then it's a trivial chage so let me know.

Reply:
I wouldn't consider this as "a problem" but my point is related to the fact that the libraries already have loose "interfaces" - which is already a bit annoying - so I think we need to be careful not to add more confusion. When I see a method `generate_r1cs_witness` or `generate_r1cs_constraints` this flips a switch and makes me think about a gadget.
Let's keep this wrapper for now, I saw your comment in the code etc. that's clear enough. We can still polish this going forward if needed.

Reply:
OK.  Sure.
(Yes.  I didn't mean to imply you had suggested forcing the inheritance - I only raised that because it was one of only a few possibilities I could see, none of which seemed like it would be the right way to go.  In any case, fine to address this in the future.)

Reply:
Sure, I'm raising this now not to forget, but as discussed this am, this is a first iteration, so more than happy to churn these gadgets later down the line :)